### PR TITLE
TW-68480 improve PE parsing & up libs

### PR DESCRIPTION
**Pull request by lesley29**
[original issue](https://youtrack.jetbrains.com/issue/TW-68480/Unity-plugin-Unity-not-detected-with-no-apparent-reason-why)

### CODE REVIEW

File: plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    fun detect(executable: File): Semver? {
+        val version = PEUtil.getProductVersion(executable)
+
+        if (version != null) {
+            return Semver("${version.p1}.${version.p2}.${version.p3}", Semver.SemverType.LOOSE)
+        }
+
+        LOG.debug("unable to detect version via PEUtil, falling back to detection tool")
+
+        val detectorPath = Paths.get("../plugins/teamcity-unity-agent/tools/PeProductVersionDetector.exe")
```
Comment (mchechulnikov) [inline plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt:26]: Maybe it will be better to extract a path and executable name to different variables and put in some place where we could reach it easily from other places, e.g. from tests. What do you think about it?

File: plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            .toAbsolutePath()
+
+        if (!detectorPath.exists()) {
+            LOG.debug("PE product version detection tool missing")
+            return null
+        }
+
+        val detector = ProcessBuilder(detectorPath.toString(), executable.absolutePath).start()
```
Comment (mchechulnikov) [inline plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt:34]: What if exception will be thrown here and along whole function?
Haven't seen try-catch here or in `WindowsUnityDetector`.

File: plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+        if (detector.waitFor(3, TimeUnit.SECONDS)) {
```
Comment (mchechulnikov) [inline plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt:36]: Could you please extract this const to named one to make searching for this timeout easier?

File: plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        }
+    }
+
+    @Serializable
+    private data class ProductVersion(
+        @SerialName("MajorPart") val majorPart: Int?,
+        @SerialName("MinorPart") val minorPart: Int?,
+        @SerialName("BuildPart") val buildPart: Int?,
+        @SerialName("PrivatePart") val privatePart: Int?
+    )
```
Comment (mchechulnikov) [inline plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt:71]: Just a question: is Unity version always semver?
If they change versioning system will the detection process be able to detect?

File: plugin-unity-agent/src/test/kotlin/jetbrains/buildServer/unity/UnityVersionsDetectorTest.kt [hunk 5]
```diff
@@ -25,7 +25,7 @@ class UnityVersionsDetectorTest {
     @Test
     fun testFindInstallations() {
         val detector = when {
-            SystemInfo.isWindows -> WindowsUnityDetector()
+            SystemInfo.isWindows -> WindowsUnityDetector(PEProductVersionDetector())
```
Comment (mchechulnikov) [inline plugin-unity-agent/src/test/kotlin/jetbrains/buildServer/unity/UnityVersionsDetectorTest.kt:28]: What do you think about adding broken (like in an original topic) binary file to check this logic in tests?

**Review by mchechulnikov — CHANGES_REQUESTED** [review]

File: plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt (hunk 4 above)
Comment (lesley29) [reply plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt:71]: It's more about the PE product version, the structure is predefined and the detection process will always work and will return numbers (or 0/null if missing).  Only in case, Unity won't put these numbers into [VS_FIXEDFILEINFO](https://learn.microsoft.com/en-us/windows/win32/api/verrsrc/ns-verrsrc-vs_fixedfileinfo) at all, the process will fail

File: plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt (hunk 2 above)
Comment (lesley29) [reply plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt:45]: Done

File: plugin-unity-agent/src/test/kotlin/jetbrains/buildServer/unity/UnityVersionsDetectorTest.kt (hunk 5 above)
Comment (lesley29) [reply plugin-unity-agent/src/test/kotlin/jetbrains/buildServer/unity/UnityVersionsDetectorTest.kt:28]: For sure, it makes sense to add some "edge case" binary. The problem is there are not many lightweight executables with specific structures. I'll investigate it later, as we don't have tests for that anyway
